lighthouse-core/lib/errors.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const strings = require('./strings');
+
+/**
+ * @typedef LighthouseErrorDefinition
+ * @property {string} code
+ * @property {string} message
+ * @property {RegExp|undefined} pattern
+ */
+
+class LighthouseError extends Error {
+  /**
+   * @param {!LighthouseErrorDefinition} errorDefinition
+   * @param {!Object=} properties
+   */
+  constructor(errorDefinition, properties) {
+    super(errorDefinition.code);
+    this.name = 'LHError';
+    this.code = errorDefinition.code;
+    this.friendlyMessage = errorDefinition.message;
+    if (properties) Object.assign(this, properties);
+
+    Error.captureStackTrace(this, LighthouseError);
+  }
+
+  /**
+   * @param {!LH.LighthouseError} err
+   */
+  static isPageLoadError(err) {
+    return err.code === errors.NO_DOCUMENT_REQUEST.code ||
+      err.code === errors.FAILED_DOCUMENT_REQUEST.code;
+  }
+
+  /**
+   * @param {string} method
+   * @param {{message: string, data: string|undefined}} protocolError
+   * @return {!Error|LighthouseError}
+   */
+  static fromProtocolMessage(method, protocolError) {
+    const matchedErrorDefinition = protocolErrors.find(e => e.pattern.test(protocolError.message));
```

Comment:
feels like doing this find() and iterating through the overall `errors` object (on L89) should be together. 

this seems fine:
```js
// extract all errors with a regex pattern to match against. findProtocolErrorDfn() then can find the matching one
const protocolErrors = Object.keys(errors).filter(k => errors[k].pattern).map(k => errors[k]);
const findProtocolErrorDfn = pattern => protocolErrors.find(e => e.pattern.test(protocolError.message));

...

const matchedErrorDefinition = findProtocolErrorDfn(protocolError.message);
```

Reply:
sg, I also ~errors~ ERRORS to make it a bit more clear it's not just a local variable, wdyt? too ugly?